articles/ros_node_idl.md

```diff
@@ -0,0 +1,205 @@
+This introduces two problems:
+
+While developers will be able to define the set of rules securing their own ROS 2 nodes, the nodes developed in-house are often outnumbered by upstream components when it comes to the entire node graph, and the developers are typically not experts in every component being used.
+Without that expertise, the entire node graph cannot be properly locked down.
+Consider a complex and popular upstream component, perhaps parts of the navigation stack.
+**Every end user** of this component must duplicate the effort of attempting to properly lock it down.
+
+If ROS 2 provided a way for upstream package authors to specify the interface required by the nodes in their package, and if the tools to generate security policies from that interface existed, neither of these problems would exist.
+
+### Other motivations
+
+Outside of security, there are several fascinating possibilities unlocked by having such an interface.
+For example, consider how this could impact [ROS 2 launch][launch_ros].
+It would be able to statically (i.e. before running anything) determine if parameter names or remappings are incorrect, among other similar sanity checks.
+Another example of the usefulness of having a static interface is the ability to create graphical tools for putting a ROS system together.
+Yet another example would be an additional feature in `ros2 pkg create` that would allow a developer to hand it an IDL and have it generate scaffolding for a node with that interface.
+
+These examples are only a subset of use-cases made possible by such an interface.
+It's clear that this is useful well beyond security.
+
+## Challenges to overcome
+
+This proposal has a number of potential upsides, but it also has some downsides worthy of discussion.
+
+### This is only really useful if it gains significant adoption in upstream packages
+
+It's true that, if not all of the packages in one's system have adopted this, its gains are incomplete.
+However, it's still useful even if only a subset of the packages adopt it (e.g. one's own packages), which means even without signifant upstream adoption it will still be useful to individuals or organizations.
+Upstream packages that haven't adopted this simply won't benefit from it.
+Also, its usefulness hopefully outweighs the work required to implement it upstream, and it's certainly something that can be contributed by community members given that the interface would be reviewed by the experts in the package.
+
+### Declared and actual interface can get out of sync
+
+This is certainly a concern: an out-of-date interface is debatably less useful than having no interface at all.
+There are a number of possibilities that will help with this issue.
+One possibility is to more tightly couple the declared and actual interface by creating a library that consumes the declared interface and creates the corresponding ROS entities.
```

Comment:
I discussed this with some people, but I don't recall who was there, perhaps @ruffsl and @thomas-moulard.  We could create an extension to the `Node` concept (much like `LifecycleNode`) that consumes the NoDL file as part of the constructor (or early in execution) and would not allow topics/services/parameters/actions outside of that policy to be created either _at all_ or past a certain point in execution (perhaps coupled with the `configure` state of `LifeCycleNode`.

Reply:
That would be an excellent way to ensure these stayed in sync. It's essentially what I was thinking when I mentioned that this could live within RCL (the next sentence, I think). The only thing I dislike about the idea is that it still requires the developers to duplicate the information: once in the NoDL and again in code. It would really be ideal if they only had to write it once, but that's a harder problem to solve.

Reply:
I think to solve that you would have to take the same approach as messages, and resort to generating base classes for node implementations to inherit from. At this point you are starting to get well into the realm of model-based tool chains and all the benefits and risks that entails. One big risk is the restrictions it places on developers in how they write and manage their code, which can become a barrier to adoption due to the increased learning required. Tools that have done this in the past have seen limited adoption while ROS has been happily adopted by people all over the place.

It's difficult to say which is the correct way to go because there are many facets to the decision.